### ISSUE TITLE
Predict narration / compare importer entries with manually edited entries

### ISSUE THREAD

**Issue by tbm** [beancount/smart_importer#37]
I haven't actually used smart_importer yet (nor have I used any beancount importer) but I've started reading the source code.

What I was hoping smart_importer would allow me to do is this: as learning input, I want to give it the raw beancount entries as generated by the importer and I want to give it a matching beancount file (i.e. same entries) after I manually edited them.

This way, payee, postings and even the narration could be figured out with machine learning.

Let's say my CVS file has:

```
Sale,04/11/2018,04/12/2018,INTERSPAR DANKT,-17.85
```

and the CVS importer will generate a transaction with `INTERSPAR DANKT` as narration.  I will then manually change this to payee `Interspar` with narration `Supermarket`.

So you'd get one beancount file with

```
2018-04-11 * "INTERSPAR DANKT"
```

and one with

```
2018-04-11 * "Interspar" "Supermarket"
```

In fact, it would contain several such entries, so smart importer could automatically figure out: oh, if the input is `INTERSPAR DANKT` we should map that to payee `Interspar` with narration `Supermarket` (and posting `Expenses:Food`).

**Comment by tbm** [comment]
Unfortunately I don't know anything about machine learning. I played around with scikit for a few hours last night and didn't quite get it to work the way I wanted, but I'll try again when I find some time.  I'll also add some links to this bug report that I found that might be of interest.

**Comment by tbm** [comment]
Someone asked about classification of ledger transactions but it's not exactly what I was looking for: https://stackoverflow.com/questions/42233232/using-scikit-learn-to-classify-multiple-outputs-of-banking-transactions

**Comment by tbm** [comment]
This looks interesting: https://towardsdatascience.com/machine-learning-nlp-text-classification-using-scikit-learn-python-and-nltk-c52b92a7c73a

**Comment by johannesjh** [comment]
> Someone asked about classification of ledger transactions but it's not exactly what I was looking for: https://stackoverflow.com/questions/42233232/using-scikit-learn-to-classify-multiple-outputs-of-banking-transactions

Interesting, thank you! I added a comment that points to this project here.

> This looks interesting: https://towardsdatascience.com/machine-learning-nlp-text-classification-using-scikit-learn-python-and-nltk-c52b92a7c73a

Yes, that's the same approach that we took (counting words and feeding them into an SVM classifier). The article details some additional stuff such as different countvectorizer options, stemming, stopwords, grid-search... each of these could be applied to the beancount smart_importers. 

In my personal experience, the predictions are fairly accurate already - I imported some transactions yesterday and they were 100% correct, I guess that was a lucky day ;-). Nonetheless, increasing the accuracy is a worthwhile goal. The article demonstrated how accuracy can be measured. So we would just need realistic test data  (ideally from multiple persons, from different accounts); then a lot of parameters can be optimized.

If you want to give it a try and try to optimize the predictions, I can recommend nteract and jupyter notebooks for that purpose; this makes it very easy to play around and try stuff out. I developed the machine learning part of `smart_importer` in this notebook:
https://gist.github.com/johannesjh/956179856957348e4fad48514b9824fc
https://nbviewer.jupyter.org/gist/johannesjh/956179856957348e4fad48514b9824fc

The notebook already contains some code where I measured the accuracy of predictions, see this line:
`print(classification_report(y_test, pred))`

**Comment by tbm** [comment]
Thanks for your comments @johannesjh. I'll look at your links later. I don't really have time to read about this at the moment but here are some other links that look promising:

* https://github.com/eli-goodfriend/banking-class
* https://techblog.commercetools.com/boosting-product-categorization-with-machine-learning-ad4dbd30b0e8 (interesting article but no code)
* https://factbased.blogspot.co.at/2012/01/categorizing-my-expenses.html
* http://jamie.works/bank-transactions.html
* https://www.digitalocean.com/community/tutorials/how-to-build-a-machine-learning-classifier-in-python-with-scikit-learn
* https://github.com/gilakos/machine-learning/blob/master/projects/capstone/archive/capstone_proposal_gilakos-V0.md
* https://towardsdatascience.com/kegra-deep-learning-on-knowledge-graphs-with-keras-98e340488b93

Also, as part of my google searches, I came across [COICOP](https://unstats.un.org/unsd/cr/registry/regcst.asp?Cl=5&Lg=1), an official classification system used for surveys. This might be an interesting resource for people trying to come up with a chart of account.  Also, if we could start a project to collect CSV bank statements, we could map to the COICOP classification manually and train a system and publish the classification (dreaming big here). Then users could use that trained data and just need a mapping between COICOP to their chart of accounts. I guess this isn't very useful for the typical beancount user, but would allow an open source personal finance system like Mint. (Again, I'm just throwing out random ideas. I'm not saying they are necessarily feasible ;).

**Comment by johannesjh** [comment]
@tbm I am sorting through the tickets and wondering what to do with this one. It contains interesting links, but what would be the specific feature, todo or next step? (I would otherwise suggest we close this ticket)

PS
Many years ago I have played around with Debian on NSLU2, and I just recognized your name and website. - So thank you for the NSLU2 installation guide!

**Comment by tbm** [comment]
Sorry @johannesjh, I haven't had any time to look into this at all and won't have time in the near future.

Let's close it for now.

Glad the NSLU2 work was useful.  Thanks for all your ledger and beancount work!
